allennlp/training/metrics/conll_coref_scores.py

```diff
@@ -0,0 +1,180 @@
+        for i, predicted_antecedent in enumerate(predicted_antecedents):
+            if predicted_antecedent < 0:
+                continue
+
+            # Find predicted index in the antecedent spans.
+            predicted_index = antecedent_indices[i, predicted_antecedent]
+            # Must be a previous span.
+            assert i > predicted_index
+            antecedent_span = tuple(top_spans[predicted_index])
+
+            # Check if we've seen the span before.
+            if antecedent_span in predicted_clusters_to_ids.keys():
+                predicted_cluster_id: int = predicted_clusters_to_ids[antecedent_span]
+            else:
+                # We start a new cluster.
+                predicted_cluster_id = len(clusters)
+                clusters.append([antecedent_span])
+                predicted_clusters_to_ids[antecedent_span] = predicted_cluster_id
+
+            mention = tuple(top_spans[i])
+            clusters[predicted_cluster_id].append(mention)
+            predicted_clusters_to_ids[mention] = predicted_cluster_id
+
+        # finalise the spans and clusters.
+        clusters = [tuple(cluster) for cluster in clusters]
+        # Return a mapping of each mention to the cluster containing it.
+        predicted_clusters_to_ids: Dict[Tuple[int, int], List[Tuple[int, int]]] = \
+            {mention: clusters[cluster_id] for mention, cluster_id
+             in predicted_clusters_to_ids.items()}
+
+        return clusters, predicted_clusters_to_ids
+
+
+class Scorer:
+    """
+    Mostly borrowed from <https://github.com/clarkkev/deep-coref/blob/master/evaluation.py>
+    """
+    def __init__(self, metric):
+        self.p_num = 0
+        self.p_den = 0
+        self.r_num = 0
+        self.r_den = 0
+        self.metric = metric
+
+    def update(self, predicted, gold, mention_to_predicted, mention_to_gold):
+        if self.metric == self.ceafe:
+            p_num, p_den, r_num, r_den = self.metric(predicted, gold)
+        else:
+            p_num, p_den = self.metric(predicted, mention_to_gold)
+            r_num, r_den = self.metric(gold, mention_to_predicted)
+        self.p_num += p_num
+        self.p_den += p_den
+        self.r_num += r_num
+        self.r_den += r_den
+
+    def get_f1(self):
+        precision = 0 if self.p_den == 0 else self.p_num / float(self.p_den)
+        recall = 0 if self.r_den == 0 else self.r_num / float(self.r_den)
+        return 0 if precision + recall == 0 else 2 * precision * recall / (precision + recall)
+
+    def get_recall(self):
+        return 0 if self.r_num == 0 else self.r_num / float(self.r_den)
+
+    def get_precision(self):
+        return 0 if self.p_num == 0 else self.p_num / float(self.p_den)
+
+    def get_prf(self):
+        return self.get_precision(), self.get_recall(), self.get_f1()
+
+    @staticmethod
+    def b_cubed(clusters, mention_to_gold):
+        """
+        <https://pdfs.semanticscholar.org/cfe3/c24695f1c14b78a5b8e95bcbd1c666140fd1.pdf>
```

Comment:
would be nice to add a sentence description for people who aren't going to click on the papers (same below as well)